src/EditorFeatures/CSharpTest/Diagnostics/GenerateType/GenerateTypeTests.cs

```diff
@@ -2018,5 +1995,58 @@ public void TestAccessibilityForPublicFields6()
 @"class A { public B<int> b = new B<int>(); public class B<T>{ public B(){}}}",
 index: 2);
         }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateType)]
+        public void TestGenerateRightMostType()
+        {
+            Test(
+@"class Program { static void Main ( string [ ] args ) { [|MyCompany|] . MyProduct . MyArea . MyClass a ; } } ",
+@"namespace MyCompany.MyProduct.MyArea { internal class MyClass { } }");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateType)]
+        public void TestGenerateRightMostTypeInsideNameOf()
+        {
+            Test(
+@"class Program { static void Main ( string [ ] args ) { var a = nameof ( [|MyCompany|] . MyProduct . MyArea . MyClass ) ; } } ",
+@"namespace MyCompany.MyProduct.MyArea { internal class MyClass { } }");
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateType)]
+        public void TestGenerateRightMostTypeNotPresentInConditional()
+        {
+            TestMissing(
+@"class Program { static void Main ( string [ ] args ) { [|MyCompany . MyProduct ? . MyArea . MyClass a |]; } }  ");
+        }
+
```

Comment:
Even if everything in `MyCompany.MyProduct?.MyArea.MyClass` bound it  would still give you errors if you put `nameof( MyCompany.MyProduct?.MyArea.MyClass)`.  Considering that there is nothing we could ever generate to make this valid I made the change to not generate types if we are a conditional access expression inside of nameof.